src/adapter/src/session.rs

```diff
@@ -695,18 +693,32 @@ impl<T: TimestampManipulation> Session<T> {
         self.vars.is_superuser()
     }
 
-    /// Returns a channel on which to send external metadata to the session.
-    pub fn retain_external_metadata_transmitter(
+    /// Register a receiver which pushes updates of [`ExternalUserMetadata`].
+    pub fn register_external_metadata_transmitter(
         &mut self,
-    ) -> UnboundedSender<ExternalUserMetadata> {
-        self.external_metadata_tx.clone()
+        rx: tokio::sync::watch::Receiver<ExternalUserMetadata>,
+    ) {
+        let old = self.external_metadata_rx.replace(rx);
+        if old.is_some() {
+            tracing::error!("Double register of external metadata transmitter!");
```

Comment:
What are the implications if this happens? Should we just panic because it means something has gone quite wrong?

Reply:
Good question! the implications are the session not getting updated external metadata. Which means a user could lose admin privileges but the session would not get updated. I turned this into an error which forces the caller to handle the scenario. We could panic but I wanted to be a bit defensive, what do you think?

Reply:
lgtm